src/sql/src/pure.rs

```diff
@@ -332,10 +331,36 @@ pub async fn purify_create_source(
                         validated_requested_subsources.push((upstream_name, subsource_name, table));
                     }
                 }
-                Some(ReferencedSubsources::Subset(subsources)) => {
-                    if publication_tables.is_empty() {
-                        sql_bail!("FOR TABLES (..) is only valid for non-empty publications");
+                Some(ReferencedSubsources::SubsetSchemas(schemas)) => {
+                    let schemas: BTreeSet<_> = schemas.iter().map(|s| s.as_str()).collect();
+                    let mut seen_schemas = BTreeSet::new();
+                    for table in &publication_tables {
+                        if !schemas.contains(table.namespace.as_str()) {
+                            continue;
+                        }
+                        seen_schemas.insert(table.namespace.as_str());
+
+                        let upstream_name = UnresolvedItemName::qualified(&[
+                            &connection.database,
+                            &table.namespace,
+                            &table.name,
+                        ]);
+                        let subsource_name = UnresolvedItemName::unqualified(&table.name);
+                        validated_requested_subsources.push((upstream_name, subsource_name, table));
                     }
+
+                    let missing_schemas: Vec<_> = schemas
+                        .difference(&seen_schemas)
+                        .map(|s| s.to_string())
+                        .collect();
+                    if !missing_schemas.is_empty() {
```

Comment:
I'm not sure if we should error based off of the list of publication tables. Given the list of published tables we can't really tell if a schema is published in full or partially and a list of publications without any table of a given schema can very well be interpreted as an empty schema, which results in a no-op.

I would suggest that we don't error here and instead check if the schema exists in the database by inspecting it during purification. If it doesn't we can error there and here we just treat valid schemas with no published tables as no-ops.